README.md
<p align="center">
  <a href="https://hub.docker.com/r/pumtrix/fnm-webui/tags?page=1&ordering=last_updated"><img src="https://img.shields.io/github/v/tag/pumtrix/fnm-webui?label=version&style=flat-square" alt="Latest Version"></a>
  <a href="https://github.com/pumtrix/fnm-webui/actions?workflow=build"><img src="https://img.shields.io/github/actions/workflow/status/pumtrix/fnm-webui/build.yml?branch=master&label=build&logo=github&style=flat-square" alt="Build Status"></a>
  <a href="https://hub.docker.com/r/pumtrix/fnm-webui/"><img src="https://img.shields.io/docker/stars/pumtrix/fnm-webui.svg?style=flat-square&logo=docker" alt="Docker Stars"></a>
  <a href="https://hub.docker.com/r/pumtrix/fnm-webui/"><img src="https://img.shields.io/docker/pulls/pumtrix/fnm-webui.svg?style=flat-square&logo=docker" alt="Docker Pulls"></a>
</p>

## About

Docker image for [FNM WebUI](https://github.com/Pumtrix-Tech/fnm-webui/), a web interface for [FastNetMon](https://fastnetmon.com).

## Special thanks

We would like to thank [@crazy-max](https://github.com/crazy-max) for his work on the LibreNMS docker repo which this is based on.

> **Note**
> 
> Want to be notified of new releases? Check out 🔔 [Diun (Docker Image Update Notifier)](https://github.com/crazy-max/diun)
> project!

___

* [Features](#features)
* [Build locally](#build-locally)
* [Image](#image)
* [Environment variables](#environment-variables)
  * [General](#general)
  * [Database](#database)
* [Volumes](#volumes)
* [Ports](#ports)
* [Usage](#usage)
  * [Docker Compose](#docker-compose)
  * [Command line](#command-line)
  * [First launch](#first-launch)
* [Upgrade](#upgrade)
* [Notes](#notes)
  * [artisan command](#artisan-command)
* [Contributing](#contributing)
* [License](#license)

## Features

* Multi-platform image
* OPCache enabled to store precompiled script bytecode in shared memory
* [s6-overlay](https://github.com/just-containers/s6-overlay/) as process supervisor
* [MariaDB](https://github.com/docker-library/mariadb) image as database instance

## Build locally

```console
$ git clone https://github.com/Pumtrix-Tech/fnm-webui-docker.git
$ cd fnm-webui-docker

# Build image and output to docker (default)
$ docker buildx bake

# Build multi-platform image
$ docker buildx bake image-all
```

## Image

Following platforms for this image are available:

```
$ docker run --rm mplatform/mquery pumtrix/fnm-webui:latest
Image: pumtrix/fnm-webui:latest
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm/v7
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
```

## Environment variables

### General

* `TZ`: The timezone assigned to the container (default `UTC`)
* `PUID`: FNM WebUI user id (default `1000`)
* `PGID`: FNM WebUI group id (default `1000`)
* `MEMORY_LIMIT`: PHP memory limit (default `256M`)
* `MAX_INPUT_VARS`: PHP max input vars (default `1000`)
* `UPLOAD_MAX_SIZE`: Upload max size (default `16M`)
* `CLEAR_ENV`: Clear environment in FPM workers (default `yes`)
* `FPM_PM_MAX_CHILDREN`: FPM max Children (default: `15`)
* `FPM_PM_START_SERVERS`: FPM start servers (default: `2`)
* `FPM_PM_MIN_SPARE_SERVERS`: FPM min spare servers (default: `1`)
* `FPM_PM_MAX_SPARE_SERVERS`: FPM max spare servers (default: `6`)
* `OPCACHE_MEM_SIZE`: PHP OpCache memory consumption (default `128`)
* `LISTEN_IPV6`: Enable IPv6 for Nginx (default `true`)
* `REAL_IP_FROM`: Trusted addresses that are known to send correct replacement addresses (default `0.0.0.0/32`)
* `REAL_IP_HEADER`: Request header field whose value will be used to replace the client address (default `X-Forwarded-For`)
* `LOG_IP_VAR`: Use another variable to retrieve the remote IP address for access [log_format](http://nginx.org/en/docs/http/ngx_http_log_module.html#log_format) on Nginx. (default `remote_addr`)
* `SESSION_DRIVER`: [Driver to use for session storage](https://github.com/Pumtrix-Tech/fnm-webui/blob/master/config/session.php) (default `file`)
* `CACHE_DRIVER`: [Driver to use for cache and locks](https://github.com/Pumtrix-Tech/fnm-webui/blob/master/config/cache.php) (default `database`)

### Database

* `DB_HOST`: MySQL database hostname / IP address
* `DB_PORT`: MySQL database port (default `3306`)
* `DB_NAME`: MySQL database name (default `fnmwebui`)
* `DB_USER`: MySQL user (default `fnmwebuis`)
* `DB_PASSWORD`: MySQL password (default `fnmwebuis`)
* `DB_TIMEOUT`: Time in seconds after which we stop trying to reach the MySQL server (useful for clusters, default `60`)

## Volumes

* `/data`: Contains configuration for future use

> **Warning**
>
> Note that the volume should be owned by the user/group with the specified
> `PUID` and `PGID`. If you don't give the volume correct permissions, the
> container may not start.

## Ports

* `8000`: HTTP port

## Usage

### Docker Compose

Docker compose is the recommended way to run this image. Copy the content of
folder [examples/compose](examples/compose) in `/var/fnm-webui/` on your host
for example. Edit the compose and env files with your preferences and run the
following commands:

```console
$ docker-compose up -d
$ docker-compose logs -f
```

### Command line

You can also use the following minimal command:

```console
$ docker run -d -p 8000:8000 --name fnm-webui \
  -v $(pwd)/data:/data \
  -e "DB_HOST=db" \
  pumtrix/fnm-webui:latest
```

> **Warning**
>
> `db` must be a running MySQL instance.

### First launch

Default user account:

username: admin@localhost
password: password

When you first access the webui, please make sure that you change the default password and update to a correct email address.

> **Note**
>
> If you lose access, you can create another one using the [`artisan` command](#artisan-command).

## Upgrade

To upgrade to the latest version of FNM WebUI, pull the newer image and launch
the container.

```console
$ docker-compose down
$ docker-compose pull
$ docker-compose up -d
```

Then run:

```console
artisan migrate
```

## Notes

### artisan command

If you want to use the `artisan` command to perform common server operations like
manage users, database migration, and more, type:

```console
$ docker-compose exec fnm-webui artisan
```

## Contributing

Want to contribute? Awesome! The most basic way to show your support is to star the project, or to raise issues. You
can also support this project by submitting pull requests to this docker repo or the main code base.

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.


CHANGELOG.md
# Changelog

## 1.0.2-r0 (2023/04/17)

* FNM WebUI 1.0.2

## 1.0.1-r0 (2023/04/17)

* FNM WebUI 1.0.1

## 1.0.0-r0 (2023/04/13)

* FNM WebUI 1.0.0


rootfs/etc/cont-init.d/04-svc-main.sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash
set -e

# From https://github.com/docker-library/mariadb/blob/master/docker-entrypoint.sh#L21-L41
# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
  local var="$1"
  local fileVar="${var}_FILE"
  local def="${2:-}"
  if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
    echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
    exit 1
  fi
  local val="$def"
  if [ "${!var:-}" ]; then
    val="${!var}"
  elif [ "${!fileVar:-}" ]; then
    val="$(<"${!fileVar}")"
  fi
  export "$var"="$val"
  unset "$fileVar"
}

DB_PORT=${DB_PORT:-3306}
DB_NAME=${DB_NAME:-fnmwebui}
DB_USER=${DB_USER:-fnmwebui}
DB_TIMEOUT=${DB_TIMEOUT:-60}

# Handle .env
if [ ! -f "/data/.env" ]; then
  echo "Generating APP_KEY and unique NODE_ID"
  cat >"/data/.env" <<EOL
APP_KEY=$(artisan key:generate --no-interaction --force --show)
NODE_ID=$(php -r "echo uniqid();")
EOL
fi
cat "/data/.env" >>"${FNMWEBUI_PATH}/.env"
chown fnmwebui:fnmwebui /data/.env "${FNMWEBUI_PATH}/.env"

file_env 'DB_PASSWORD'
if [ -z "$DB_PASSWORD" ]; then
  echo >&2 "ERROR: Either DB_PASSWORD or DB_PASSWORD_FILE must be defined"
  exit 1
fi

dbcmd="mysql -h ${DB_HOST} -P ${DB_PORT} -u "${DB_USER}" "-p${DB_PASSWORD}""
unset DB_PASSWORD

echo "Waiting ${DB_TIMEOUT}s for database to be ready..."
counter=1
while ! ${dbcmd} -e "show databases;" >/dev/null 2>&1; do
  sleep 1
  counter=$((counter + 1))
  if [ ${counter} -gt ${DB_TIMEOUT} ]; then
    echo >&2 "ERROR: Failed to connect to database on $DB_HOST"
    exit 1
  fi
done
echo "Database ready!"

# Enable first run wizard if db is empty
counttables=$(echo 'SHOW TABLES' | ${dbcmd} "$DB_NAME" | wc -l)
echo ${dbcmd}
echo "Count: " ${counttables}
if [ "${counttables}" -eq "0" ]; then
  echo "Enabling First Run Wizard..."
  echo "INSTALL=user,finish">> ${FNMWEBUI_PATH}/.env
fi

echo "Updating database schema..."
artisan migrate --force --no-ansi --no-interaction
artisan db:seed --force --no-ansi --no-interaction

echo "Clear cache"
artisan cache:clear --no-interaction
artisan config:cache --no-interaction

mkdir -p /etc/services.d/nginx
cat >/etc/services.d/nginx/run <<EOL
#!/usr/bin/execlineb -P
with-contenv
s6-setuidgid ${PUID}:${PGID}
nginx -g "daemon off;"
EOL
chmod +x /etc/services.d/nginx/run

mkdir -p /etc/services.d/php-fpm
cat >/etc/services.d/php-fpm/run <<EOL
#!/usr/bin/execlineb -P
with-contenv
s6-setuidgid ${PUID}:${PGID}
php-fpm81 -F
EOL
chmod +x /etc/services.d/php-fpm/run



rootfs/etc/cont-init.d/03-config.sh
#!/usr/bin/with-contenv bash
# shellcheck shell=bash
set -e

# From https://github.com/docker-library/mariadb/blob/master/docker-entrypoint.sh#L21-L41
# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
  local var="$1"
  local fileVar="${var}_FILE"
  local def="${2:-}"
  if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
    echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
    exit 1
  fi
  local val="$def"
  if [ "${!var:-}" ]; then
    val="${!var}"
  elif [ "${!fileVar:-}" ]; then
    val="$(<"${!fileVar}")"
  fi
  export "$var"="$val"
  unset "$fileVar"
}

TZ=${TZ:-UTC}

MEMORY_LIMIT=${MEMORY_LIMIT:-256M}
UPLOAD_MAX_SIZE=${UPLOAD_MAX_SIZE:-16M}
CLEAR_ENV=${CLEAR_ENV:-yes}
FPM_PM_MAX_CHILDREN=${FPM_PM_MAX_CHILDREN:-15}
FPM_PM_START_SERVERS=${FPM_PM_START_SERVERS:-2}
FPM_PM_MIN_SPARE_SERVERS=${FPM_PM_MIN_SPARE_SERVERS:-1}
FPM_PM_MAX_SPARE_SERVERS=${FPM_PM_MAX_SPARE_SERVERS:-6}
OPCACHE_MEM_SIZE=${OPCACHE_MEM_SIZE:-128}
LISTEN_IPV6=${LISTEN_IPV6:-true}
REAL_IP_FROM=${REAL_IP_FROM:-"0.0.0.0/32"}
REAL_IP_HEADER=${REAL_IP_HEADER:-"X-Forwarded-For"}
LOG_IP_VAR=${LOG_IP_VAR:-remote_addr}
MAX_INPUT_VARS=${MAX_INPUT_VARS:-1000}

MEMCACHED_PORT=${MEMCACHED_PORT:-11211}

DB_PORT=${DB_PORT:-3306}
DB_NAME=${DB_NAME:-fnmwebui}
DB_USER=${DB_USER:-fnmwebui}
DB_TIMEOUT=${DB_TIMEOUT:-30}

FNMWEBUI_BASE_URL=${FNMWEBUI_BASE_URL:-/}

# Timezone
echo "Setting timezone to ${TZ}..."
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} >/etc/timezone

# PHP
echo "Setting PHP-FPM configuration..."
sed -e "s/@MEMORY_LIMIT@/$MEMORY_LIMIT/g" \
  -e "s/@UPLOAD_MAX_SIZE@/$UPLOAD_MAX_SIZE/g" \
  -e "s/@CLEAR_ENV@/$CLEAR_ENV/g" \
  -e "s/@FPM_PM_MAX_CHILDREN@/$FPM_PM_MAX_CHILDREN/g" \
  -e "s/@FPM_PM_START_SERVERS@/$FPM_PM_START_SERVERS/g" \
  -e "s/@FPM_PM_MIN_SPARE_SERVERS@/$FPM_PM_MIN_SPARE_SERVERS/g" \
  -e "s/@FPM_PM_MAX_SPARE_SERVERS@/$FPM_PM_MAX_SPARE_SERVERS/g" \
  /tpls/etc/php81/php-fpm.d/www.conf >/etc/php81/php-fpm.d/www.conf

echo "Setting PHP INI configuration..."
sed -i "s|memory_limit.*|memory_limit = ${MEMORY_LIMIT}|g" /etc/php81/php.ini
sed -i "s|;date\.timezone.*|date\.timezone = ${TZ}|g" /etc/php81/php.ini
sed -i "s|;max_input_vars.*|max_input_vars = ${MAX_INPUT_VARS}|g" /etc/php81/php.ini

# OpCache
echo "Setting OpCache configuration..."
sed -e "s/@OPCACHE_MEM_SIZE@/$OPCACHE_MEM_SIZE/g" \
  /tpls/etc/php81/conf.d/opcache.ini >/etc/php81/conf.d/opcache.ini

# Nginx
echo "Setting Nginx configuration..."
sed -e "s#@UPLOAD_MAX_SIZE@#$UPLOAD_MAX_SIZE#g" \
  -e "s#@REAL_IP_FROM@#$REAL_IP_FROM#g" \
  -e "s#@REAL_IP_HEADER@#$REAL_IP_HEADER#g" \
  -e "s#@LOG_IP_VAR@#$LOG_IP_VAR#g" \
  /tpls/etc/nginx/nginx.conf >/etc/nginx/nginx.conf

if [ "$LISTEN_IPV6" != "true" ]; then
  sed -e '/listen \[::\]:/d' -i /etc/nginx/nginx.conf
fi

rm -rf ${FNMWEBUI_PATH}/logs
rm -f ${FNMWEBUI_PATH}/config.d/*
mkdir -p /etc/logrotate.d
touch /etc/logrotate.d/fnmwebui

echo "Setting FNM WebUI configuration..."

# Env file
if [ -z "$DB_HOST" ]; then
  echo >&2 "ERROR: DB_HOST must be defined"
  exit 1
fi
file_env 'DB_PASSWORD'
if [ -z "$DB_PASSWORD" ]; then
  echo >&2 "ERROR: Either DB_PASSWORD or DB_PASSWORD_FILE must be defined"
  exit 1
fi
cat >${FNMWEBUI_PATH}/.env <<EOL
APP_URL=${FNMWEBUI_BASE_URL}
DB_HOST=${DB_HOST}
DB_PORT=${DB_PORT}
DB_DATABASE=${DB_NAME}
DB_USERNAME=${DB_USER}
DB_PASSWORD=${DB_PASSWORD}
EOL

ln -sf /data/logs ${FNMWEBUI_PATH}/logs

# Fix perms
echo "Fixing perms..."
chown -R fnmwebui:fnmwebui ${FNMWEBUI_PATH}/bootstrap ${FNMWEBUI_PATH}/storage
#chown -R fnmwebui:fnmwebui /data/logs ${FNMWEBUI_PATH}/bootstrap ${FNMWEBUI_PATH}/logs ${FNMWEBUI_PATH}/storage
#chmod ug+rw /data/logs ${FNMWEBUI_PATH}/bootstrap/cache ${FNMWEBUI_PATH}/storage ${FNMWEBUI_PATH}/storage/framework/*
chmod ug+rw ${FNMWEBUI_PATH}/bootstrap/cache ${FNMWEBUI_PATH}/storage ${FNMWEBUI_PATH}/storage/framework/*
